hm_pyhelper/miner_param.py

```diff
@@ -76,25 +64,96 @@ def run_gateway_mfr(args):
         raise ECCMalfunctionException(err_str).with_traceback(e.__traceback__)
 
 
+def get_gateway_mfr_path() -> str:
+    direct_path = os.path.dirname(os.path.abspath(__file__))
+    return os.path.join(direct_path, 'gateway_mfr')
+
+
+def get_gateway_mfr_version() -> Version:
+    """
+    Returns version of gateway_mfr
+    """
+    gateway_mfr_path = get_gateway_mfr_path()
+    command = [gateway_mfr_path, '--version']
+
+    try:
+        run_gateway_mfr_result = subprocess.run(
+            command,
+            capture_output=True,
+            check=True
+        )
+    except Exception as e:
+        err_str = f"Exception occurred on running gateway_mfr --version: {e}"
+        LOGGER.exception(err_str)
+        raise GatewayMFRExecutionException(err_str).with_traceback(e.__traceback__)
+
+    # Parse gateway_mfr version
+    try:
+        version_str = run_gateway_mfr_result.stdout.rpartition(' ')[-1]
+        gateway_mfr_version = Version(version_str)
+
+        return gateway_mfr_version
+    except Exception as e:
+        err_str = f"Exception occurred while parsing gateway_mfr version: {e}"
+        LOGGER.exception(err_str)
+        raise GatewayMFRInvalidVersion(err_str).with_traceback(e.__traceback__)
+
+
+def get_gateway_mfr_command(sub_command: str) -> list:
+    gateway_mfr_path = get_gateway_mfr_path()
+    command = [gateway_mfr_path]
+
+    gateway_mfr_version = get_gateway_mfr_version()
+    if gateway_mfr_version.minor == 1:
+        try:
+            device_arg = [
+                '--path',
+                get_variant_attribute(os.getenv('VARIANT'), 'KEY_STORAGE_BUS')
+            ]
+            command.extend(device_arg)
+        except (UnknownVariantException, UnknownVariantAttributeException) as e:
+            LOGGER.warning(str(e) + ' Omitting --path arg.')
+
+        command.append(sub_command)
+
+        # In case of "key" command, append the slot number 0 at the end.
+        if sub_command == "key":
+            command.append("0")
+
+    elif gateway_mfr_version.minor == 2:
+        try:
+            device_arg = [
+                '--device',
+                get_variant_attribute(os.getenv('VARIANT'), 'SWARM_KEY_URI')
+            ]
+            command.extend(device_arg)
+        except (UnknownVariantException, UnknownVariantAttributeException) as e:
+            LOGGER.warning(str(e) + ' Omitting --device arg.')
+
+        command.append(sub_command)
```

Comment:
Do we need an `else` case if `minor` is not 1 or 2?

Reply:
Thanks for good catch. I would like to raise an exception when `minor` is not 1 or 2, 'UnsupportedGatewayMfrVersion` and test cases for it.

Reply:
Or it would be better to fall back to using v0.2.1 format. That is, new ECC addressing specification will be used for the v0.2.1 and upward.

Reply:
The code for version check is updated. For the version >= 0.2.0 `--device` parameter is used, for the version from `0.1.1` to `o.1.9`, legacy `--path` parameter is used. For the other versions(invalid versions led to `0.0.0`), raises an exception called `UnsupportedGatewayMfrVersion`.